### PR TITLE
FP selection, arity validation in Algorithm constructor, no-replacement option in tournament selection

### PR DESCRIPTION
**Pull request by itaitzruia4**
- Implemented Fitness Proportionate selection (AKA Roulette Wheel selection)
- Added arity validation in Algorithm (to prevent runtime errors in cases of 10 population size and 3 parents crossover for instance)
- Added `replace` parameter in Tournament selection, that allows the choice of individuals for the tournament to be either with or without replacement (by default, replace=True for backwards compatibility)

### CODE REVIEW

File: eckity/algorithms/algorithm.py [hunk 1]
```diff
@@ -151,6 +151,22 @@ def __init__(
...
+            selection_methods = sub_pop.get_selection_methods()
+            selection_methods = [t[0] for t in selection_methods]
+
+            operators_sequence = sub_pop.get_operators_sequence()
+            operators = selection_methods + operators_sequence
+
+            for oper in operators:
+                if sub_pop.population_size % oper.arity != 0:
+                    raise ValueError(
+                        f"Operator {oper} arity must be "
```
Comment (achiyae) [inline eckity/algorithms/algorithm.py:167]: Move to breeder

File: eckity/algorithms/tests/test_algorithm.py
Comment (achiyae) [inline eckity/algorithms/tests/test_algorithm.py:1]: should be moved to breeder as well

File: eckity/genetic_operators/selections/fp_selection.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @override
+    def select(self, source_inds, dest_inds):
+        n_selected = len(source_inds) - len(dest_inds)
+
+        fitness_scores = np.array(
+            [ind.get_augmented_fitness() for ind in source_inds]
+        )
+        # convert higher fitness scores to be better
+        if not self.higher_is_better:
+            fitness_scores = -fitness_scores
```
Comment (achiyae) [inline eckity/genetic_operators/selections/fp_selection.py:42]: need to check that there are no minus fitness values. see https://github.com/GMUEClab/ecj/blob/10a5debd0edbb47fb0fba02da79c8086db9e1dc9/ecj/src/main/java/ec/select/FitProportionateSelection.java#L71

**Review by achiyae — CHANGES_REQUESTED** [review]

**Comment by itaitzruia4** [comment]
@achiyae all changes were done, version-detector still doesn't work properly
I think we can just remove it and manually check version changes in CR

**Pull request merged**